### PR TITLE
feat: organize with Older and prerelease packages

### PR DESCRIPTION
**Pull request by burkedavison**

### CODE REVIEW

File: third_party/docfx-doclet-143274/src/test/java/com/microsoft/util/ApiVersionTest.java [hunk 1]
```diff
@@ -70,4 +73,37 @@ public void equals() {
   private ApiVersion parse(String s) {
     return ApiVersion.parse(s).orElseThrow(() -> new IllegalStateException("Unable to parse " + s));
   }
+
+  @Test
+  public void testRecommendation_PrioritizesReleaseVersions() {
+    ImmutableList<ApiVersion> versions =
+        ImmutableList.of(parse("v101beta"), parse("v2p1"), parse("v1p14alpha15"));
```
Comment (alicejli) [inline third_party/docfx-doclet-143274/src/test/java/com/microsoft/util/ApiVersionTest.java:80]: Consider adding a test case where there are 2 release versions for completeness (maybe 2 release versions and one prerelease or something).

File: third_party/docfx-doclet-143274/src/main/java/com/microsoft/build/YmlFilesBuilder.java [hunk 2]
```diff
@@ -19,16 +22,18 @@
...
-  private String outputPath;
-  private ElementUtil elementUtil;
-  private PackageLookup packageLookup;
-  private String projectName;
-  private boolean disableChangelog;
-  private ProjectBuilder projectBuilder;
-  private PackageBuilder packageBuilder;
-  private ClassBuilder classBuilder;
-  private ReferenceBuilder referenceBuilder;
+  private static final String OLDER_AND_PRERELEASE = "Older and prerelease packages";
```
Comment (alicejli) [inline third_party/docfx-doclet-143274/src/main/java/com/microsoft/build/YmlFilesBuilder.java:25]: (out of ignorance) - what's the purpose of making these `final`?

File: third_party/docfx-doclet-143274/src/main/java/com/microsoft/lookup/PackageLookup.java [hunk 3]
```diff
@@ -51,4 +66,112 @@ public String extractJavaType(PackageElement element) {
...
+   * This 'grouping' logic combines all versioned packages together in a single `a.b.c.v#` group.
+   *
+   * <p>For example: a.b.v1 and a.b.v2 will be in the same group, but a.b and a.b.c will be in their
+   * own groups.
+   *
+   * <p>When packages are grouped, only one package within the group will be VISIBLE and the rest
+   * will be placed in the OLDER_AND_PRERELEASE category.
+   */
+  @VisibleForTesting
+  Multimap<String, PackageElement> groupVersions(List<PackageElement> packages) {
```
Comment (alicejli) [inline third_party/docfx-doclet-143274/src/main/java/com/microsoft/lookup/PackageLookup.java:130]: IIUC, does this mean that there is one recommended package per stable version? So if an API has v1, v1alpha, v2, v2beta1, then v1 and v2 would be visible and v1alpha and v2beta1 would be hidden under the "Older and Prerelease packages"?

**Review by alicejli — COMMENTED** [review]
Some clarifying questions on how this works; thanks for your work on this! If possible to add mocked staging just to double-check this renders properly that would be helpful as well.

File: third_party/docfx-doclet-143274/src/main/java/com/microsoft/build/YmlFilesBuilder.java (hunk 2 above)
Comment (alicejli) [reply third_party/docfx-doclet-143274/src/main/java/com/microsoft/build/YmlFilesBuilder.java:25]: Discussed offline; making fields `final` is preferred if possible in general as it helps the compiler optimize and prevents unwanted behavior (assuming that having the field be unchanged is indeed the desired behavior).

File: third_party/docfx-doclet-143274/src/main/java/com/microsoft/lookup/PackageLookup.java (hunk 3 above)
Comment (burkedavison) [reply third_party/docfx-doclet-143274/src/main/java/com/microsoft/lookup/PackageLookup.java:130]: One version will be visible, and the rest will be hidden under "Older and prerelease packages". 
Ex: If v1 and v2 exist, only v2 is visible and v1 is under "Older..."
Ex: If v1 and v2beta exists, only v1 is visible and v2beta is under "Older and prerelease..."

File: third_party/docfx-doclet-143274/src/test/java/com/microsoft/util/ApiVersionTest.java (hunk 1 above)
Comment (burkedavison) [reply third_party/docfx-doclet-143274/src/test/java/com/microsoft/util/ApiVersionTest.java]: Done

**Comment by burkedavison** [comment]
Staging complete, reviewed, and discussed.

**Review by alicejli — APPROVED** [review]
Looks good, thank you! Putting this into a comment for future reference:

1) Follow up needed on handling stub packages
2) Current implementation selects a single package to highlight. If there is need in the future for either a manual override or for multiple packages to be highlighted, that will be a new FR.

**Pull request merged**